state/machine_test.go

```diff
@@ -336,6 +336,59 @@ func (s *MachineSuite) TestLifeJobManageModel(c *gc.C) {
 	c.Assert(err, gc.ErrorMatches, "machine 0 is still a voting controller member")
 }
 
+func (s *MachineSuite) TestLifeJobManageModelWithControllerCharm(c *gc.C) {
+	cons := constraints.Value{
+		Mem: newUint64(100),
+	}
+	changes, err := s.State.EnableHA(3, cons, "quantal", nil)
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(changes.Added, gc.HasLen, 2)
+
+	m2, err := s.State.Machine("2")
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(m2.Jobs(), gc.DeepEquals, []state.MachineJob{
+		state.JobHostUnits,
+		state.JobManageModel,
+	})
+
+	ch := s.AddTestingCharmWithSeries(c, "juju-controller", "")
+	app := s.AddTestingApplicationForSeries(c, "quantal", "controller", ch)
+	unit, err := app.AddUnit(state.AddUnitParams{})
+	c.Assert(err, jc.ErrorIsNil)
+	err = unit.SetCharmURL(ch.URL())
+	c.Assert(err, jc.ErrorIsNil)
+	err = unit.AssignToMachine(m2)
+	c.Assert(err, jc.ErrorIsNil)
+	err = m2.Destroy()
+	c.Assert(err, jc.ErrorIsNil)
+
+	c.Assert(m2.Life(), gc.Equals, state.Alive)
+	c.Assert(err, jc.ErrorIsNil)
+
+	for {
+		needsCleanup, err := s.State.NeedsCleanup()
```

Comment:
normally we'd assert the expected cleanup count
at the least can we assert that NeedsCleanup() is true before running the for loop